enhancements/builds/multi-arch-builds.md

```diff
@@ -0,0 +1,456 @@
+
+### Goals
+
+* Provide users of Builds with a minimal — in cases where builds are based on images we provide, preferably zero — effort path to having their images ready to run on any node in their multi-architecture clusters.
+
+### Non-Goals
+
+* Building images for architectures for which the cluster has no running nodes.
+* Building multi-arch images which depend on base images which are not multi-arch.
+
+## Proposal
+
+### openshift-controller-manager
+
+Instead of creating one (1) pod to perform a build followed by an optional
+push, create N+1 pods when building for N architectures.
+
+The first N pods will build an ordinary image for their respective targets and,
+if the build spec specifies an output location, push their images and update a
+platform-specific build status field in the build resource.
+
+The last pod will use functionality to be added to the openshift builder to
+build a manifest list and push it to the output location.
+
+### openshift-builder
+
+The builder binary is currently invoked by multiple names to perform multiple
+tasks in build pods.
+
+We'll extend openshift-builder's `openshift-docker-build` and
+`openshift-sti-build` commands to perform the build as they already do, with
+two exceptions: if the build spec specifies a target platform, the build will
+target that platform for the build, and the build status will be written to the
+corresponding platform-specific status field in the Build object which owns the
+build pod.
+
+We'll extend openshift-builder by adding another command it can be invoked by
+which will tell it to monitor the status of the other builder pods. If the
+builder pods all complete successfully, and the build spec specifies an output
+location, it will build and push a manifest list using the information gathered
+from the build status information. As its final act, it will attempt to untag
+the individual images from the registry.
+
+Note: untagging should be friendlier to registries which perform garbage
+collection, particularly for cases where the build fails on one or more
+architectures, and as a result a manifest list is not pushed. It should also
+reduce "noise" in the list of tags reported by registries which support listing
+the tags in a repository.
+
+### Workflow Description
+
+#### Variation 1/3: SCM integrated
+1. Application developer merges a PR to their repository's main branch.
+1. The SCM platform notifies OpenShift via a webhook.
+1. OpenShift has previously associated the webhook with triggering a BuildConfig.
+
+#### Variation 2/3: not SCM integrated
+1. Application developer manually triggers a BuildConfig.
+
+#### Variation 3/3: not SCM integrated
+1. Application developer creates a Build.
+
+#### Common
+1. The openshift-controller-manager starts N+1 build pods.
+1. Platform-specific images are pushed to the registry.
```

Comment:
> Why would you imagine that to be best? 

to ensure uniqueness and traceability

> Would the build instance id contain the specific architecture as well? 

i wouldn't expect it to automatically, no.  i'd just expect it to look like "build-foo-$uuid" or "build-foo-$counter"

>   I would imagine that it would be simplest to just tag it with <desired-tag>-<arch> and the manifest list would be tagged with <desired-tag>.

including the build uuid in the tag certainly isn't any more difficult.  Whether it's useful/necessary?  perhaps i was overzealous, but keep in mind that in traditional builds today, the user picked the tag to push to.  With this new pattern, we are not (at least per the api defined so far) allowing the user to choose the tags we're going to push the arch specific images to.  So the more we do to ensure they won't collide with something, the better.  Running multiple builds in parallel (i.e. the whole multiarch build getting kicked off in parallel, not just running arch specific builds in parallel) is also a consideration for uniqueness.

Reply:
Users would be responsible for ensuring that tags are unique and non-colliding today. In this new scenario, that tag would likely be used by the manifestlist. Therefore, non-uniqueness of a manifestlist would result in the expected behavior of non-uniqueness of component architecture builds in my opinion. Instead, we would only be interested in ensuring that the images for different architectures do not collide and amending with the architecture flavor should be enough here. 

As a user, I would expect to get greater value and comprehension from an architecture modifier than a build id modifier.

It seems like we don't need to be overly protective here as we are not trying to enforce an NVR uniqueness. One complication is the behavior around if a user doesn't want a manifestlist. If we always generate one and instead just surface the arch-specific image(s), we need to make sure that we can appropriately tag those images according to the requested tag. I imagine that this only needs to be backwards compatible if one architecture is requested without a manifest list and not if multiple architectures are requested.

Reply:
> In this new scenario, that tag would likely be used by the manifestlist. 

no, the manifestlist needs to use SHAs, not tags, for references.  the "intermediate tags" are, at least as the EP currently stands, an implementation detail largely hidden from the user and only used to allow the images to be pushed to the registry and not garbage collected until the underlying SHAs can be stitched into the manifestlist.

> As a user, I would expect to get greater value and comprehension from an architecture modifier than a build id modifier.

one doesn't preclude also having the other, though.